docs/design/task-launch-order-within-job.md

```diff
@@ -0,0 +1,104 @@
+| matlab     | 2        | linear dependencies | yes    | 1. Add initcontainer to the task<br/>2.Using user-written check scripts in initcontainer |1.Increase the cost of use for users<br />2.Still using resources while waiting|
+| tensorflow | n>=2   | linear dependencies | yes    | 1. Add initcontainer to the task<br/>2.Using user-written check scripts in initcontainer |1.Increase the cost of use for users<br />2.Still using resources while waiting|
+
+## Requirement
+
+Based on the scenarios listed above, the dependencies can be abstracted as:
+
+* task-A depends on task-B, which means A must be started first and then B.
+* Triggering policy, for example level triggering and edge triggering when handling circuit. In our cases, the triggering policy may be the task's event, like task running/restarted/completed. 
+
+For the ease of end user's experince, we need to unify the way of composing a complicated jobs with lots of tasks, instead of letting user handle the complexcity themselvs using init-containers or other workflow tools. So we need to  have an more advanced VCJob that has below abilities:
+
+* Each task can have one or more tasks it depends on, and it can specify the triggering policy.
+* Examples of triggering policys can be:
+  * task starts to run
+  * task restarts
+  * task completes
+* User may specify the time period after the triggering event happens, for example
+  * Imediately after task completed
+  * Half an hour afater task starts to run or completes
+* Besides the time period, user may specify an certain condition as the triggering policy, for example,
+  * Check if a specified file exists in the /data directory.
+  * Check if the specified port is open, such as 2222.
+  * Check if the HTTP GET request to the specified path was successfully accessed, such as /health.
+  
+## Design
+
+### API
+taskflow example
+```
+apiVersion: batch.volcano.sh/v1alpha1
+kind: TaskFlow
+metadata:
+  name: task-flow
+spec:
+  tasks:
+  - name: A
+  - name: B
+    depends: 
+      target: ['A']
+      check: liveness
+      action: ['ssh', 'root@1.1.1.1']
+  - name: C
+    depends: 
+      target: ['B']
+      check: readiness
+      action: []
+  - name: D
+    depends:
+      target: ['B']
+      check: status
+      action: compeleted
+  - name: E
+    depends:
+      target: ['C', 'D']
+      check: command
+      action: ['ls', '/data']
+      iteration: any/all
+```
+
+### Usage
+* create taskflow `task-flow`
+* create volcano job `vc-job`, spec.taskflow is the taskflow name, make sure task names corresponding to taskflow
+* get task status to check if it is correct order
+
+### Implementaion
+* generate new CRD **TaskFlow**
+* create admission for TaskFlow, VCJob adds new fields
+* overwrite `allocate` action, name as a new action
+
+### Notice
+* deal with the conflict with gang-scheduling
```

Comment:
This should also be clarified, how to avoid conflicts with gang scheduling. And if we need to overwrite `allocate`, what's the basic workflow?

Reply:
abount gang, we discussed two options: 
1. task level gang scheduling
2. taskflow and gang do not coexist
abount allocate: 
1. create a new action
2. create a new component

We can discuss the above options specifically in our weekly meetings.